IanGRap/Squish
# AI First Steps

Contributors: Emma, Coulter

Goal: 
- Create a basic "King of the Hill" AI prototype as a proof of concept before attempting to actually make AI opponents a full feature. Basic AI functionality should include dashing, jumping, avoiding overhead blocks, and trying to reach the highest point in the arena.

Files Changed:
- AIBlockSearch.cs: Script that implements the King of the Hill algorithm, as initially proposed by Emma. Uses delegate to tell AIInput.cs what input should be activated. Inputs are dash, jump, horizontal movement, and vertical movement. If AI is within a certain distance from a block, it will jump onto it. If it is within a slightly larger distance, it will dash. Gizmos have also been added so that in the Scene view in Unity, one can see a visual representation of the AI player's dash and jump ranges.
- AIInput.cs: Sets and stores values for various input variables based on information received from the delegate in AIBlockSearch.cs.
- PlayerInput.cs: Modified to receive AI input from AIInput.cs for movement, dashing and jumping, and controls whether AI input is enabled based on specific key presses. Keys to enable AI: "7" for player 1, "8" for player 2.
- PlayerData.cs: Added "IsAIPlayer" public variable so that each player object can keep track of whether or not its AI input has been enabled. This variable is in PlayerData.cs so that the value of this variable will persist between rounds, so if a player's AI input is enabled halfway through the game, it will continue to be enabled until either the appropriate key is pressed or the game ends completely.
- BlockDamager.cs: Added Height field for use in calculating how tall a column of blocks is, so that an AI player's ability to reach the top can be determined.
- Player.prefab: Set dash and jump distance values to 2.3 and 1.75, respectively, based on testing different values to determine what was within the maximum distance a player can dash and jump.
- One By Two Block.prefab: Set Height value to 2.
- Tall Block.prefab: Set Height value to 3.
- AI Test.unity: Test scene for AI functionality.

How To Test:
- Start the Lobby scene. 
- Set game mode to "Classic". Spawn players 1 and 2. 
- Enter the tutorial. 
- Press "7" and/or "8" to activate AI input for the corresponding player. 
- Use keyboard or controllers to move the players close to blocks and observe their behavior when near blocks. They should attempt to jump or dash onto nearby blocks depending on their distance from the top of a block column (observe the gizmos in the scene view to be sure).
- Move the players to the platform to start a game.
- Observe their behavior during play. They should be always trying to escape blocks above them, and will jump or dash on top of blocks/ columns next to them if they're close enough.
- Finish game and do a rematch.
- Press "7" and/or "8" multiple times during play to ensure that AI input is activated or deactivated as intended. The state of input enabled/disabled should persist between rounds.
- Go through the win screen and back through the lobby and another match to ensure that nothing breaks.

Notes (PLEASE READ):
- There is a lot of wonkiness to the AI's behavior. Currently, it will repeatedly make decisions that don't necessarily seem logical, and may even result in getting the AI player killed. This is simply meant to be a very first prototype to ensure that creating an AI that acts and reacts based on an algorithm is possible. The algorithm will need much more complex behavior and decision-making ability added to it before the AI can actually be expected to make intelligent decisions. This AI just needs to react in simple ways to simple calculations. If it can be observed to do that consistently, then this prototype is a success.
